homie-controller/src/types.rs

```diff
@@ -394,4 +478,283 @@ mod tests {
             Err(ParseExtensionError("test.blah:1.2.3:4.x".to_owned()))
         );
     }
+
+    #[test]
+    fn property_integer_parse() {
+        let mut property = Property::new("property_id");
+
+        // With no known value, parsing fails.
+        assert_eq!(property.value::<i64>(), Err(ValueError::Unknown));
+
+        // With an invalid value, parsing also fails.
+        property.value = Some("-".to_owned());
+        assert_eq!(
+            property.value::<i64>(),
+            Err(ValueError::ParseFailed {
+                value: "-".to_owned(),
+                datatype: Datatype::Integer,
+            })
+        );
+
+        // With a valid value but unknown datatype, parsing succeeds.
+        property.value = Some("42".to_owned());
+        assert_eq!(property.value(), Ok(42));
+
+        // With the correct datatype, parsing still succeeds.
+        property.datatype = Some(Datatype::Integer);
+        assert_eq!(property.value(), Ok(42));
```

Comment:
I have a strong preference for one test per function, so that you can work out what's going on when you make a bunch of them fail. I would suggest:

* Extract each of the comments to be test names.
* potentially use nested `mod`s to help with grouping of tests
* There are a few ways to make instantiating structs easier (make a `prop("42")` helper, and do `Property { datatype: Some(Datatype::Integer), ..prop("42") }`

https://crates.io/crates/test-case will also help, if you want something that's more like table driven testing.

Reply:
... that said, I can hardly dictate the style of tests in the project if I refuse to write any, so maybe merge it and I'll show you what I mean when I finally write some tests.

Reply:
You mean one test per assert? That seems a bit over the top I think. In this case I quiet like modifying the property and checking that I get the results I expect each step of the way.

Reply:
I found it a bit difficult to follow which of your mutations were still in flight and interesting vs which had been shadowed vs which were in flight but not interesting to the current assertion.

Splitting tests doesn't have to be much more verbose if you have easy ways to construct the various states, and it can be much easier to follow and debug.

A comment thread is probably not the best place to discuss this. Let's either pair on it, or merge it and I will have a go at refactoring at a later date.